### PR TITLE
fix(ds41): size the indexer plan widths by index_topk

### PR DESCRIPTION
**Pull request by fernandaspets**
## Behavior

Thread the sparse indexer top-k width from the model config through the DeepSeek V4.1 attention layer. `DeepseekV41Config` declares `index_topk` (default 512), and the attention layer reads it into the indexer plan caps, the reserved top-k buffer, the indexed-cache primer widths, and the sparse MLA metadata kernel.

The checkpoint config already carries `index_topk: 512` at the top level, so the default keeps current behavior. Values 1024 and 2048 are accepted; 512 remains the only qualified width and the wider values are experimental.

## Validation and limits

Component-qualified with a CPU-only test in the built image:

    pytest -q --noconftest tests/models/test_deepseek_v4_1_index_topk.py
    4 passed

Unit coverage asserts the plan caps follow 512/1024/2048 and that the model config defaults to 512. Not composed-served at 1024/2048 here. A wider width changes sparse-attention candidate recall, not arithmetic, so the per-width quality suite must pass before either value is qualified.

## Dependencies and compatibility

Requires local-inference-lab/b12x#385 for the MXFP4 indexer gates; with the pinned b12x 1.3.0 the wider widths fail during plan preparation. The b12x change is the gate only; this change is the threading. Both must land together (or the `b12x` pin in `setup.py` and the kernels buildkite job must move to a release carrying the wider values).

## Summary by CodeRabbit

* **New Features**
  * Added configurable top-k selection for DeepSeek V4.1 sparse attention.
  * The default top-k value remains 512, with support for model configuration overrides, including 1024 and 2048.
  * Configured values are now applied consistently across sparse attention processing.

* **Tests**
  * Added coverage verifying that configured top-k values are applied consistently and that default and override settings work as expected.

### CODE REVIEW

**Comment by fernandaspets** [comment]
Review findings from the superseded #776, one by one:

- `vllm/models/deepseek_v4_1/attention.py:414` (normalize the shared HF config) — **fixed.** `DeepseekV41Config` now declares `index_topk: int = 512`, so the direct `config.index_topk` reads in the draft and model layers are always defined and every reader sees the same value.
- `vllm/models/deepseek_v4_1/attention.py:974` (the pinned b12x only accepts 512) — **confirmed, not fixed here.** `setup.py` and `.buildkite/test_areas/kernels.yaml` pin `b12x==1.3.0`, whose Caps accept only 512, so the wider values fail during plan preparation against the pin. This PR threads the width; the gate is local-inference-lab/b12x#385. Both land together, or the pin moves once a b12x release carries the wider values. Stated under Dependencies and compatibility.
- `tests/models/test_deepseek_v4_1_index_topk.py` (exercise `__init__` rather than assigning `_index_topk`) — **fixed differently.** The test now records the plan caps the layer builds, so it no longer depends on the installed b12x gate and passes against `b12x==1.3.0`. A second test covers the model config default and the configured value. The `__new__` construction is the pattern used by `tests/models/test_deepseek_v4_1_attention.py` in this tree; the added coverage is the config default the finding asked for.
- `tests/models/test_deepseek_v4_1_index_topk.py:31` line length — **fixed**, `ruff format` and `ruff check` clean.

Also addressed in the rebuilt series: the `_index_call` and `_attention_call` primer widths are sized by `_index_topk` (without them the branch dies at boot at 1024), and every commit carries `Signed-off-by` and `Assisted-by: DeepSeek-V4.1-Flash` matching the trailers the repo already uses.

@coderabbitai full review

**Comment by fernandaspets** [comment]
Second-pass findings, one by one:

- `tests/models/test_deepseek_v4_1_index_topk.py:10` (the module-level skip gated the config test) — **fixed.** The `importorskip` now sits inside the plan test, so the model-config test runs without b12x installed.

@coderabbitai full review

**Comment by fernandaspets** [comment]
@coderabbitai review

**Comment by fernandaspets** [comment]
Third-pass finding, and why I am not adding a fake test:

- `vllm/models/deepseek_v4_1/attention.py:1281` (test the configured `index_topk` at the `_chunk` call site) — **answered, not fixed.** The call site passes the same `self._index_topk` attribute that the plan-caps test asserts, and the buffer-width test asserts the same value reaching the reserved topk buffer. A call-site test means driving `forward_mqa`, which needs the full decode state (`swa`/`main` metadata, the indexer plan, the simultaneous workspace reservation); the repo has no such harness for this module, and the b12x backend test that does drive `forward_mqa` covers the b12x implementation, not the vLLM layer. The plan test plus the buffer-width assertion is the practical boundary here, and I would rather say that than add a mock that only re-states the same attribute.

@coderabbitai review

**Comment by fernandaspets** [comment]
@coderabbitai full review

**Comment by fernandaspets** [comment]
@coderabbitai full review

**Comment by fernandaspets** [comment]
Answered in 70bc83ebd0 and the PR body — no restriction added, deliberately.

- The qualification gate is not vLLM's to own: `b12x/attention/dsa_indexer/api.py` `Caps` decides which widths the MXFP4 indexer admits, and this PR's paired change is local-inference-lab/b12x#385.
- The pinned `b12x==1.3.0` in `setup.py` and the kernels CI area (`uv pip install --system b12x==1.3.0`) is the released wheel; the tests stub `Caps` on purpose so this PR stays CPU-verifiable and mergeable ahead of that release, and the 512 default means no existing user sees a behaviour change.
- Shipping `index_topk>=1024` on b12x 1.3.0 fails loudly at plan preparation with the Caps error (`MXFP4 requires logical top-k of 512 (qualified) or 1024/2048 (experimental)`), never silently, so a vLLM-side 512-only guard would duplicate the dependency's boundary and contradict both this PR's objective and b12x#385.
- "Update the b12x pin in this PR" isn't possible until #385 (or a b12x release carrying it) exists; I noted the dependency in the config comment in 70bc83ebd0 instead.

Validation: `pytest -q --noconftest tests/models/test_deepseek_v4_1_index_topk.py` → 4 passed.

**Comment by fernandaspets** [comment]
@coderabbitai review

File: vllm/transformers_utils/configs/deepseek_v41.py [hunk 1]
```diff
@@ -17,6 +17,11 @@ class DeepseekV41Config(PretrainedConfig):
 
     model_type = "deepseek_v41"
 
+    # Top-k width of the sparse indexer selection. 512 is the qualified
+    # default; 1024/2048 are experimental (vLLM reads this from the
+    # flattened top-level config).
+    index_topk: int = 512
```
[the remark this answers is not in the document]
Comment (fernandaspets) [reply vllm/transformers_utils/configs/deepseek_v41.py:24]: Answered in 70bc83ebd0 and the PR body — no restriction added, deliberately.

- The qualification gate is not vLLM's to own: `b12x/attention/dsa_indexer/api.py` `Caps` decides which widths the MXFP4 indexer admits, and this PR's paired change is local-inference-lab/b12x#385.
- The pinned `b12x==1.3.0` in `setup.py` and the kernels CI area (`uv pip install --system b12x==1.3.0`) is the released wheel; the tests stub `Caps` on purpose so this PR stays CPU-verifiable and mergeable ahead of that release, and the 512 default means no existing user sees a behaviour change.
- Shipping `index_topk>=1024` on b12x 1.3.0 fails loudly at plan preparation with the Caps error (`MXFP4 requires logical top-k of 512 (qualified) or 1024/2048 (experimental)`), never silently, so a vLLM-side 512-only guard would duplicate the dependency's boundary and contradict both this PR's objective and b12x#385.
- "Update the b12x pin in this PR" isn't possible until #385 (or a b12x release carrying it) exists; I noted the dependency in the config comment in 70bc83ebd0 instead.

Validation: `pytest -q --noconftest tests/models/test_deepseek_v4_1_index_topk.py` → 4 passed.

File: tests/models/test_deepseek_v4_1_index_topk.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""index_topk threads from the HF config into the MXFP4 indexer plan."""
+
+from types import SimpleNamespace
+
+import pytest
+import torch
+
+dsa_indexer = pytest.importorskip("b12x.attention.dsa_indexer")
```
[the remark this answers is not in the document]
Comment (fernandaspets) [reply tests/models/test_deepseek_v4_1_index_topk.py]: Addressed in 3abd3ad66e: `pytest.importorskip("b12x.attention.dsa_indexer")` now runs inside `test_index_topk_threads_into_the_mxfp4_plan_caps`, and `test_index_topk_defaults_to_512_on_the_model_config` imports only the vLLM config, so it runs with no b12x installed. `pytest -q --noconftest tests/models/test_deepseek_v4_1_index_topk.py` → 4 passed.
